packages/kit/src/runtime/client/router.js

```diff
@@ -6,6 +8,25 @@ function scroll_state() {
 		y: pageYOffset
 	};
 }
+/**
+ * @param {URL} url
+ * @returns {boolean}
+ */
+function dispatch_navigation_intent(url) {
+	let allow_navigation = true;
+
+	dispatchEvent(
+		new CustomEvent('sveltekit:navigation-intent', {
+			detail: {
+				url,
+				cancel: () => {
+					allow_navigation = false;
+				}
+			}
+		})
+	);
+	return allow_navigation;
+}
```

Comment:
If I understand correctly, the `sveltekit:navigation-intent` event is sync only. So if the user wants to do something async in `onBeforeNavigate`, it's not possible? I believe async code would be a common usecase that we'd want to support

Reply:
@bluwy Thanks for the detailed review! 

I thought about that when I was creating the navigator and came to the conclusion that I should not make it async.

Three reasons for  that:
- CustomEvents are sync only, there is no way for me to wait until the `sveltekit:navigation-intent` returns something. And it is becoming even trickier if we have multiple `onBeforeNavigation` intents defined on one page.  Now we need to make sure that every intent was completed correctly before we navigate. 

- Making it async might even create a bad UX for the user since he will be wondering what is happening and why it he not navigating anywhere.

- There is a workaround that I've created for async tasks.
If we have to make a complex check or an API request before leaving the page we could simply:
1) Save the URL we are going to navigate to in memory or in global storage
2) Prevent navigation from happening
3) Show a Loading Spinner to the user
4) Once a task completes we use goto(savedURL) to manually leave the page after the async task is completed.
4.1) The programmatic goto() will trigger the `onBeforeNavigate` interceptor a second time but it is up to the user to create the condition that will prevent the heavy async from firing.

Example async implementation
```
let savedUrl
let heavyStuffDone = false;

onBeforeNavigate((navigationIntent) => {
        if(heavyStuffDone) return;

	savedUrl = navigationIntent.url;
	navigationIntent.cancel();

        const canNavigate = await DoSomethingHeavy()
        
        if(canNavigate){
            heavyStuffDone = true;
            goto(savedUrl.href)
        }
});
```

Reply:
> Example async implementation
> 
> ```
> 	let heavyStuffDone = false;
> 
> 	onBeforeNavigate(async (navigationIntent) => {
> 		if (heavyStuffDone) return;
> 		navigationIntent.cancel();
> 
> 		await new Promise((r) => setTimeout(r, 2000));
> 
> 		const canNavigate = true;
> 		if (canNavigate) {
> 			heavyStuffDone = true;
> 			goto(navigationIntent.url.pathname);
> 		}
> 	});
> ```

Pushing a history entry to the browser history is suboptimal when the user tried to navigate backwards.

Reply:
> The CustomEvent is not really needed, is it?
I saw you pr form onNavigate in sapper, I like some of the ideas, I will try to remove the CustomEvent, the main issue with CustomEvent is that we cannot have multiple callbacks that talk to the parent, cannot have it truly async and have HMR issues in dev in vite. 

> Most of the time, you want to display a popup that informs the user that there are unsaved changes in a form and ask the user if they still want to navigate or not.
Yeah it is true, I was trying to say that it is always expected to show a modal/loading spinner even during heavy async stuff before navigating, meaning we would never need to wait for async stuff to complete to show something, we always show stuff before heavy stuff happens.

>Pushing a history entry to the browser history is suboptimal when the user tried to navigate backward.
This is correct, and I will fix you as you suggested in you other comments to write the history entry index into every history state and navigate back to the current history entry, meaning once this is done the async example with the custom event will still a valid option for async stuff, but I agree that it may look like a hack

Reply:
Thanks for the explanation @zommerberg. I still think preventing async isn't the most ergonomic API though. Having it async, the function should still be able to control the loading state before & after the `await`, so UX-wise I don't think it's an issue. At least for me, I think having `async` stall the navigation would match my expectation of it, instead of recommending another `goto`.

Since you mentioned about removing `CustomEvent`s, perhaps async could work again?

Reply:
Yes, I after think about it I agree, we should remove CustomEvent and make it async

Reply:
What would be the most ideal approach here?

Reply:
I like the idea on how @PatrickG  did it in https://github.com/PatrickG/sapper-navigation-enhancer

Reply:
Awesome, I think there's a few ways to tackle this. I haven't looked too deep into @PatrickG's implementation, but having some sort of list of functions that are passed to `onNavigate` might work, and then running them sequentially.

Reply:
```js
async trigger_on_before_navigate_callbacks(url) {
		const callbacksNavigationResult = !(
			await Promise.all(this.on_before_navigate_callbacks.map(([callback]) => callback(url)))
		).some((result) => result === false);

		return callbacksNavigationResult;
	}
```	

This is the direction I am heading to, the onBeforeNavigate callback needs to return a boolean

Reply:
And internally this function will be called during goto,click and popstate events:

```js
const allow_navigation = this.trigger_on_before_navigate_callbacks(url);
if (!allow_navigation) return;
```